### ISSUE TITLE
discussion: add `in` operator as sugar for `has_edge`

### ISSUE THREAD

**Issue by sbromberger** [JuliaGraphs/LightGraphs.jl#748]
`e in g` as sugar for `has_edge` has a nice simplicity to it. We decided against this early on because it wasn't clear that edges were the best/only thing that was an element of a graph, but I'm open to revisiting (hat tip @mschauer). 

This would also give us `e ∈ g` and `e ∉ g`, which I think look nice and are intuitive. However, the fact that we won't have `for e in g` as an iterator might be confusing.

Thoughts?

cc @jpfairbanks @juliohm @somil55 @mschauer

**Comment by mschauer** [comment]
A data point
```
m0019971:~/.julia/v0.6/LightGraphs$ grep -r "has_edge" . | wc -l
      64
m0019971:~/.julia/v0.6/LightGraphs$ grep -r "has_vertex" . | wc -l
       9
```
also `v in vertices(g)` is performant and reads relatively nicely.

**Comment by sbromberger** [comment]
> also `v in vertices(g)` is performant and reads relatively nicely.

We also use `e in edges(g)` right now though.

**Comment by mschauer** [comment]
Isn't that much slower?
```
julia> @btime has_edge($g, 1,2)
  35.602 ns (0 allocations: 0 bytes)

julia> @btime $e in edges($g)
  4.068 μs (1 allocation: 32 bytes)
```

**Comment by juliohm** [comment]
I personally don't find this sugar syntax very appealing. Mathematically, can we say an edge belongs to a graph in a set sense? I've never seen this in the graph literature, please let me know if it is common. My vote is to leave has_edge explicit as it is now.

**Comment by sbromberger** [comment]
Proof of concept:
```
julia> in(e::AbstractEdge, g::AbstractGraph) = has_edge(g, e)
in (generic function with 41 methods)

julia> g = PathGraph(10)
{10, 9} undirected simple Int64 graph

julia> in(Edge(1,2), g)
true

julia> Edge(1,2) in g
true

julia> Edge(1,3) in g
false

julia> Edge(1,2) ∈ g
true

julia> Edge(1,3) ∈ g
false

julia> Edge(1,3) ∉ g
true
```

**Comment by sbromberger** [comment]
> Isn't that much slower?

Well, yes, because it's not `has_edge` - it's an iterator over all edges:

```
for e in edges(g)
  # do something with e
end
```

**Comment by sbromberger** [comment]
> My vote is to leave has_edge explicit as it is now.

@juliohm and others : please register your vote by thumbs-up or thumbs-down on the original post. It will make it easier to tally. Thanks.

**Comment by mschauer** [comment]
@juliohm The plan is keep LightGraphs using explicit `has_edge`. The discussion is whether to allow it as syntactical sugar.

**Comment by sbromberger** [comment]
> Mathematically, can we say an edge belongs to a graph in a set sense?

The literature commonly includes

Consider a graph $G = \{V, E\}$ comprised of a set of vertices $V$ and a set of edges $E \subseteq V \times V$ between vertices.

**Comment by juliohm** [comment]
Exactly, according to this definition there is no such thing as "graph set interface", thanks for raising this discussion, my vote is thumbs down for this one.

**Comment by mschauer** [comment]
Mathematicians constantly speak of "... if the edge is in the graph...", only that usually E is defined to denote edges and `e in E` is used unlike Julia where `E` has nothing to do with `G`.

**Comment by sbromberger** [comment]
> From a formal point of view, a graph is an ordered pair <V,E> with two sets. Operations of "has element" apply to the constituent sets, not the graph. So, I would find counter-intuitive to have the `\in` unicode applied directly to graphs.

This was precisely the reason we elected not to do it the last time around, IIRC.

**Comment by juliohm** [comment]
I agree @mschauer , we sometimes get lazy and take shortcuts to talk math. However, just because we can be informal speaking, it doesn't mean we can be informal while coding :blush: This sugar seems to add something else to think about that is not math, and that doesn't improve day-to-day usage of LG.

**Comment by sbromberger** [comment]
Thank you all for the discussion and feedback. I'm going to close this out since it doesn't look like it enjoys overwhelming support.

**Comment by sbromberger** [comment]
Actually, what about this proposal?

```
in(e::AbstractEdge, i::AbstractEdgeIter) = LightGraphs.insorted(dst(e), i.adj[src(e)])
```
gives us O(log(deg)) membership testing via `e in edges(g)`, where I think we can argue that `edges(g)` is a (mathematical) set.

**Comment by somil55** [comment]
@sbromberger I think it will good idea to provide this because in discrete mathematics we consider graph to be a set of vertices and edges.

**Comment by mschauer** [comment]
With #751 just
```
in(e::Edge, i::SimpleEdgeIter) = has_edge(i.g, e)
```

**Comment by jpfairbanks** [comment]
Yeah this works the edge iterator is an (ordered) set of edges so membership makes sense.

I agree that the in a strict sense V and E are "in" G=(V,E). So the only two things in a graph should be vertices(g) and edges(g).

**Comment by sbromberger** [comment]
@mschauer do you want to file a PR based on https://github.com/JuliaGraphs/LightGraphs.jl/issues/748#issuecomment-328767422? I'm about to merge #751.

**Comment by mschauer** [comment]
Thanks to the opposition (and some work) the story in the end is much nicer than the initial sketch. 👍

**Comment by sbromberger** [comment]
Closed via #752. Thanks, @mschauer :)
